### ISSUE TITLE
[java] FP: `UnnecessaryReturn` flags a `return;` that exits a loop early

### ISSUE THREAD

**Issue by Chordrain** [pmd/pmd#6746]
**PMD Version:** 7.24.0
**Category:** [UnnecessaryReturn](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#unnecessaryreturn)
**Severity:** Warning (Code Style)

## Summary

PMD's `UnnecessaryReturn` rule is meant to flag a `return;` whose removal would not change behavior — i.e. a redundant `return;` at the textual end of a void method.

The rule fires on a `return;` that sits inside a nested `for` loop and serves to **exit the loop (and method) early**. Removing that `return;` is *not* behavior-preserving: the loop would continue iterating. Flagging it as "unnecessary" is therefore a **false positive**.

### before — `throw` inside the loop (no finding)

```java
public class LoopExitBefore {
    void m(Object[] os) {
        int y = 0;
        for (Object o : os) {
            for (int j = 0; j < 10; j++) {
                y += o.hashCode();
                if ((y & 1) == 0)
                    throw new RuntimeException();
            }
        }
    }
}
```

### after — early-exit `return;` (false positive)

```java
public class LoopExitAfter {
    void m(Object[] os) {
        int y = 0;
        for (Object o : os) {
            for (int j = 0; j < 10; j++) {
                y += o.hashCode();
                if ((y & 1) == 0)
                    return;   // FALSE POSITIVE: exits both loops early
            }
        }
    }
}
```

## Expected behavior

`UnnecessaryReturn` should not fire on the AFTER `return;`. Removing it changes control flow: the inner `for (int j = 0; j < 10; j++)` would keep iterating for the remaining `j` values, and in subsequent outer iterations for every remaining `o`. The "this return has no effect" precondition is not met.

## Actual behavior

- `throw new RuntimeException();` inside the loop → 0 findings (correct).
- `return;` inside the loop → 1 finding: the `return;` is reported as unnecessary, even though it exits the loop early.

The matcher appears to treat any `return;` that has no later sibling statement in its immediate block as redundant, missing the case where the enclosing block is a loop body whose continued iterations would produce observable behavior.

**Comment by UncleOwen** [comment]
Can't reproduce:

```
$ cat LoopExitAfter.java 
public class LoopExitAfter {
    void m(Object[] os) {
        int y = 0;
        for (Object o : os) {
            for (int j = 0; j < 10; j++) {
                y += o.hashCode();
                if ((y & 1) == 0)
                    return;   // FALSE POSITIVE: exits both loops early
            }
        }
    }
}
$ pmd-7.24.0 check --rulesets=category/java/codestyle.xml/UnnecessaryReturn --no-progress --no-cache --dir=.
$
```

What are you doing differently?

**Comment by Chordrain** [comment]
I apologize for the noise. After a closer re-examination, I realized I made a mistake in my initial verification, and this issue actually cannot be reproduced.

It was a false alarm on my end. I will close this issue now. Sorry for any inconvenience caused!
